src/basic/path-util.c

```diff
@@ -639,8 +639,58 @@ static int check_x_access(const char *path, int *ret_fd) {
         return 0;
 }
 
-int find_executable_full(const char *name, const char *root, bool use_path_envvar, char **ret_filename, int *ret_fd) {
-        int last_error, r;
+static bool find_executable_from_path(const char *name, const char *root, const char *element, int *last_error_addr, char **ret_filename, int *ret_fd, int fd) {
+        int r;
+        _cleanup_free_ char *dup_element = strdup(element);
+
+        assert(element);
+
+        if (!path_is_absolute(dup_element))
+                return false;
+
+        if (!path_extend(&dup_element, name))
```

Comment:
hmm, sounds weird first dupping the path and then extending it. why not instead just call path_join() to get the joined result in one step?

Reply:
yes, strdup() creates a copy of a string. that's ok. but you then extend the string with path_extend() in a second step. which internally realloc()s the original string if necessary.

instead: just use path_join() which concats all passed strings and places the result in a newly allocated string too, but does htis in one step. This is both shorter and faster.